decentraland/unity-explorer
# feat: support local gltf v2

Tasks:

- [x] Apply correct GameObject Layer on processed colliders based on the `invisibleMeshesCollisionMask` and new mesh colliders based on the `visibleMeshesCollisionMask` properties of the `GLTFContainer` that comes from the SDK. A good test scene to compare how it should work (only for `visibleMeshesCollisionMask`) is [Goerli Plaza's block-dog scene](https://github.com/decentraland/sdk7-goerli-plaza/blob/047ab35cd9ae4ab389774a3275b93361769185e8/Block-dog/src/dog.ts#L26~L30). When loading asset bundles this process is already working correctly, so the raw GLTF colliders should just enter the same pipeline. [Other goerli scenes using those properties](https://github.com/search?q=repo%3Adecentraland%2Fsdk7-goerli-plaza%20invisibleMeshesCollisionMask%3A%20ColliderLayer.&type=code).
- [x] Implement animation processing to achieve parity with assets from assetbundles (Currently animations don't work as expected) 
>(8/9/2024 edit: Animations are working now, we need to check parity with the new support for layers introduced recently in E@)
>(9/19/2024 edit: Animations blends are supported as long they don't affect the same bones, this might be shaped in the future, reach out to Aga or Alejandro Melucci for more details)
- [x] Implement our own MaterialBuilder or make sure we are using the same material we use for loading all GltfContainers with Asset Bundles. Provide it to GltfImporter on construction.
- [x] Make sure that if the Explorer is not running in preview / “local scene development” mode, the raw GLTF loading systems shouldn't run. That detection is currently at the main [local scene development with E@ PR](https://github.com/decentraland/unity-explorer/pull/1392).

The work continues in v3 #2075 

-----------------------------------------

# QA TEST INSTRUCTIONS

The testing consists of opening 2 instances of the Explorer Alpha on a specific scene/world and compare between the two if everything **looks** the same or at least almost the same.

## Setup

### Local Scene (uses raw GLTFs)

1. Download [this test scene](https://github.com/decentraland/sdk7-test-scenes/tree/main/scenes/54%2C-55-Testing-3d-models). Either by cloning the repo or downloading the compressed repo, just **open a terminal inside that folder**.
2. run `npm i @dcl/sdk@latest`
3. run `npm run start -- --explorer-alpha` -> that will open the launcher and latest released E@. **CLOSE THEM**, the test has to be done with the build from this PR, not from the launcher. The scene server will keep running in the terminal, don't kill it.
4. Open the build downloaded from this PR connecting it to the running local scene. 
    * WinOS example (terminal in same folder as the build) -> `"C:\Users\PravsDesktop\Downloads\Decentraland_windows64\Decentraland.exe" --realm http://127.0.0.1:8000 --position 54,-55 --local-scene true`. You can also do that by creating a direct link to the build and in the "Target" field of its properties putting the `--realm http://127.0.0.1:8000 --position 54,-55 --local-scene true --debug` part and oening the direct link file.
    * MacOS example (terminal in same folder as the build) -> `open Decentraland.app -n --args --realm http://127.0.0.1:8000 --position 54,-55 --local-scene true --debug`
5. You should see the scene that contains several walls with lots of GLTFs
6. In the debug panel, expand the SKYBOX section and click on "Pause", set the time to `40000` and click on "Set Time"

### World Scene (uses Asset Bundle GLTFs)

1. Open the build downloaded from this PR connecting it to deployed scene at the `nullrefxeption` world. 
    * WinOS example (terminal in same folder as the build) -> `"C:\Users\PravsDesktop\Downloads\Decentraland_windows64\Decentraland.exe" --realm http://127.0.0.1:8000 --position 54,-55 --local-scene true`. You can also do that by creating a direct link to the build and in the "Target" field of its properties putting the `--realm nullrefxeption.dcl.eth --position 54,-55 --debug` part and oening the direct link file.
    * MacOS example (terminal in same folder as the build) -> `open Decentraland.app -n --args --realm nullrefxeption.dcl.eth --position 54,-55 --debug`
2. You should see the scene that contains several walls with lots of GLTFs
3. In the debug panel, expand the SKYBOX section and click on "Pause", set the time to `40000` and click on "Set Time"

## TESTING

The idea is to open both Explorers in parallel and compare how everything looks, paying particular attention to these parts of the scene, confirming that they look the same or almost the same:

**Alpha Blend**
<img width="1194" alt="Screenshot 2024-09-20 at 5 32 31 PM" src="https://github.com/user-attachments/assets/0ee209d0-d5a0-4c48-8dc5-c8f6c6c242db">

<img width="1328" alt="Screenshot 2024-09-20 at 5 32 49 PM" src="https://github.com/user-attachments/assets/71a0087e-a909-4813-9eb5-a1d1a417d886">

**Emissive materials**

<img width="1154" alt="Screenshot 2024-09-20 at 5 32 39 PM" src="https://github.com/user-attachments/assets/43689dae-6954-4bf4-b360-62ab92d8f99e">

**Specular** 

<img width="1312" alt="Screenshot 2024-09-20 at 5 05 02 PM" src="https://github.com/user-attachments/assets/2336b3ee-c240-475b-bd1e-e6c94a40541e">
(Ignore the label "AssetBundle" in the screenshot)
